From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sat, 14 Nov 2020 09:36:43 -0800
Subject: [PATCH] Fixed kick event leave message not being sent


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index a736d676ef5e3a25c144743af559bfb46e294db7..f003ce2093ca7cd73f230cf31301cdd9675befd8 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -304,7 +304,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
         this.networkManager.sendPacket(new PacketPlayOutKickDisconnect(ichatbasecomponent), (future) -> {
             this.networkManager.close(ichatbasecomponent);
         });
-        this.a(ichatbasecomponent); // CraftBukkit - fire quit instantly
+        this.onAllDisconnect(ichatbasecomponent, event.getLeaveMessage()); // CraftBukkit - fire quit instantly // Paper
         this.networkManager.stopReading();
         MinecraftServer minecraftserver = this.minecraftServer;
         NetworkManager networkmanager = this.networkManager;
@@ -1659,6 +1659,12 @@ public class PlayerConnection implements PacketListenerPlayIn {
 
     @Override
     public void a(IChatBaseComponent ichatbasecomponent) {
+        // Paper start
+        this.onAllDisconnect(ichatbasecomponent, this.minecraftServer.getPlayerList().disconnect(this.player, "\u00A7e" + this.player.getName() + " left the game"));
+    }
+
+    private void onAllDisconnect(IChatBaseComponent ichatbasecomponent, String quitMessage) {
+        // Paper end
         // CraftBukkit start - Rarely it would send a disconnect line twice
         if (this.processedDisconnect) {
             return;
@@ -1674,7 +1680,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
         */
 
         this.player.p();
-        String quitMessage = this.minecraftServer.getPlayerList().disconnect(this.player);
+        // String quitMessage = this.minecraftServer.getPlayerList().disconnect(this.player); // Paper
         if ((quitMessage != null) && (quitMessage.length() > 0)) {
             this.minecraftServer.getPlayerList().sendMessage(CraftChatMessage.fromString(quitMessage));
         }
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 1faefcb5d23e2b86932611ba850fb7b60de896c7..1b57dd0eae4e94ea88db91091bdc6abf015ff4f6 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -498,7 +498,7 @@ public abstract class PlayerList {
 
     }
 
-    public String disconnect(EntityPlayer entityplayer) { // CraftBukkit - return string
+    public String disconnect(EntityPlayer entityplayer, String leaveMessage) { // CraftBukkit - return string // Paper - add default disconnect message
         WorldServer worldserver = entityplayer.getWorldServer();
 
         entityplayer.a(StatisticList.LEAVE_GAME);
@@ -509,7 +509,7 @@ public abstract class PlayerList {
             entityplayer.closeInventory(org.bukkit.event.inventory.InventoryCloseEvent.Reason.DISCONNECT); // Paper
         }
 
-        PlayerQuitEvent playerQuitEvent = new PlayerQuitEvent(cserver.getPlayer(entityplayer), "\u00A7e" + entityplayer.getName() + " left the game", entityplayer.quitReason); // Paper - quit reason
+        PlayerQuitEvent playerQuitEvent = new PlayerQuitEvent(cserver.getPlayer(entityplayer), leaveMessage, entityplayer.quitReason); // Paper - quit reason
         if (entityplayer.didPlayerJoinEvent) cserver.getPluginManager().callEvent(playerQuitEvent); // Paper - if we disconnected before join ever fired, don't fire quit
         entityplayer.getBukkitEntity().disconnect(playerQuitEvent.getQuitMessage());
 
