From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chase <chasewhip20@gmail.com>
Date: Wed, 2 Dec 2020 22:43:39 -0800
Subject: [PATCH] add per world spawn limits

Taken from #2982. Credit to Chasewhip8

diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 5596d9f42523dbcbad47d21a4b2fdba39e3c5590..c3be91536494e8bebff111e25d76f6ec07fb370c 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -616,6 +616,19 @@ public class PaperWorldConfig {
         zombieVillagerInfectionChance = getDouble("zombie-villager-infection-chance", zombieVillagerInfectionChance);
     }
 
+    public int spawnLimitMonsters = -1;
+    public int spawnLimitAnimals = -1;
+    public int spawnLimitWaterAnimals = -1;
+    public int spawnLimitWaterAmbient = -1;
+    public int spawnLimitAmbient = -1;
+    private void perWorldSpawnLimits() {
+        spawnLimitMonsters = getInt("spawn-limits.monsters", spawnLimitMonsters);
+        spawnLimitAnimals = getInt("spawn-limits.animals", spawnLimitAnimals);
+        spawnLimitWaterAnimals = getInt("spawn-limits.water-animals", spawnLimitWaterAnimals);
+        spawnLimitWaterAmbient = getInt("spawn-limits.water-ambient", spawnLimitWaterAmbient);
+        spawnLimitAmbient = getInt("spawn-limits.ambient", spawnLimitAmbient);
+    }
+
     public int lightQueueSize = 20;
     private void lightQueueSize() {
         lightQueueSize = getInt("light-queue-size", lightQueueSize);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index b71316cce3bdbf3485be456f0260c6b3463cff8e..a101a5dde3f665e009659755fb9b810404146b0b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -341,6 +341,13 @@ public class CraftWorld implements World {
         this.generator = gen;
 
         environment = env;
+        // Paper start - per world spawn limits
+        this.monsterSpawn = this.world.paperConfig.spawnLimitMonsters;
+        this.animalSpawn = this.world.paperConfig.spawnLimitAnimals;
+        this.waterAnimalSpawn = this.world.paperConfig.spawnLimitWaterAnimals;
+        this.waterAmbientSpawn = this.world.paperConfig.spawnLimitWaterAmbient;
+        this.ambientSpawn = this.world.paperConfig.spawnLimitAmbient;
+        // Paper end
     }
 
     @Override
