From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Thu, 3 Dec 2020 00:05:32 -0800
Subject: [PATCH] Make hoppers respect inventory max stack size


diff --git a/src/main/java/net/minecraft/server/TileEntityHopper.java b/src/main/java/net/minecraft/server/TileEntityHopper.java
index d432de40eba2767f4ced4d9c642c9d2033acd0ea..ff1a0b806ac60dee1efd672bb6ce3d032328f6d6 100644
--- a/src/main/java/net/minecraft/server/TileEntityHopper.java
+++ b/src/main/java/net/minecraft/server/TileEntityHopper.java
@@ -569,13 +569,21 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
             boolean flag1 = iinventory1.isEmpty();
 
             if (itemstack1.isEmpty()) {
+                // Paper start
+                int origCount = itemstack.getCount();
+                itemstack.setCount(Math.min(iinventory1.getMaxStackSize(), itemstack.getCount()));
                 IGNORE_TILE_UPDATES = true; // Paper
-                iinventory1.setItem(i, itemstack);
+                iinventory1.setItem(i, itemstack.cloneItemStack());
                 IGNORE_TILE_UPDATES = false; // Paper
+                if (origCount - itemstack.getCount() == 0) {
                 itemstack = ItemStack.b;
+                } else {
+                    itemstack.setCount(origCount - itemstack.getCount());
+                }
+                // Paper end
                 flag = true;
             } else if (a(itemstack1, itemstack)) {
-                int j = itemstack.getMaxStackSize() - itemstack1.getCount();
+                int j = Math.min(itemstack.getMaxStackSize(), iinventory1.getMaxStackSize()) - itemstack1.getCount(); // Paper
                 int k = Math.min(itemstack.getCount(), j);
 
                 itemstack.subtract(k);
